File: Algorithms/src/1292. Maximum Side Length of a Square with Sum Less than or Equal to Threshold/Code/solution.go
func maxSideLength(mat [][]int, threshold int) int {
    m, n := len(mat), len(mat[0])
    pre := make([][]int, m+1)

    for i := 0; i <= m; i++ {
        pre[i] = make([]int, n+1)
    }

    // Prefix sum
    for i := 1; i <= m; i++ {
        for j := 1; j <= n; j++ {
            pre[i][j] = mat[i-1][j-1] +
                pre[i-1][j] +
                pre[i][j-1] -
                pre[i-1][j-1]
        }
    }

    left, right := 0, min(m, n)
    ans := 0

    for left <= right {
        mid := (left + right) / 2
        found := false

        for i := mid; i <= m && !found; i++ {
            for j := mid; j <= n; j++ {
                sum := pre[i][j] -
                    pre[i-mid][j] -
                    pre[i][j-mid] +
                    pre[i-mid][j-mid]

                if sum <= threshold {
                    found = true
                    break
                }
            }
        }

        if found {
            ans = mid
            left = mid + 1
        } else {
            right = mid - 1
        }
    }

    return ans
}

func min(a, b int) int {
    if a < b {
        return a
    }
    return b
}

File: Algorithms/src/1292. Maximum Side Length of a Square with Sum Less than or Equal to Threshold/Code/solution.py
class Solution:
    def maxSideLength(self, mat, threshold):
        m, n = len(mat), len(mat[0])

        # Prefix sum matrix
        pre = [[0] * (n + 1) for _ in range(m + 1)]

        for i in range(1, m + 1):
            for j in range(1, n + 1):
                pre[i][j] = mat[i - 1][j - 1] \
                            + pre[i - 1][j] \
                            + pre[i][j - 1] \
                            - pre[i - 1][j - 1]

        left, right = 0, min(m, n)
        ans = 0

        while left <= right:
            mid = (left + right) // 2
            found = False

            for i in range(mid, m + 1):
                for j in range(mid, n + 1):
                    square_sum = pre[i][j] \
                               - pre[i - mid][j] \
                               - pre[i][j - mid] \
                               + pre[i - mid][j - mid]

                    if square_sum <= threshold:
                        found = True
                        break
                if found:
                    break

            if found:
                ans = mid
                left = mid + 1
            else:
                right = mid - 1

        return ans
